protocols/directdefund/directdefund.go

```diff
@@ -0,0 +1,189 @@
+
+// errors
+var ErrNotApproved = errors.New("objective not approved")
+var ErrChannelUpdateInProgress = errors.New("cannot defund channel with unsupported, non-final latest state")
+
+// Objective is a cache of data computed by reading from the store. It stores (potentially) infinite data
+type Objective struct {
+	Status protocols.ObjectiveStatus
+	C      *channel.Channel
+}
+
+func isUpdateInProgress(c *channel.Channel) bool {
+	latestSS, err := c.LatestSignedState()
+	// There are no signed states
+	if err != nil {
+		return false
+	}
+	return !latestSS.HasAllSignatures() && !latestSS.State().IsFinal
+}
+
+// NewObjective initiates an Objective with the supplied channel
+func NewObjective(
+	preApprove bool,
+	c *channel.Channel,
+) (Objective, error) {
+	// We are chosing to disallow creating an objective if the channel has an in-progress update
+	if isUpdateInProgress(c) {
+		return Objective{}, ErrChannelUpdateInProgress
+	}
+
+	var init = Objective{}
+
+	if preApprove {
+		init.Status = protocols.Approved
+	} else {
+		init.Status = protocols.Unapproved
+	}
+	init.C = c.Clone()
+
+	return init, nil
+}
+
+// Public methods on the DirectDefundingObjective
+func (o Objective) Id() protocols.ObjectiveId {
+	return protocols.ObjectiveId(ObjectivePrefix + o.C.Id.String())
+}
+
+func (o Objective) Approve() Objective {
+	updated := o.clone()
+	// todo: consider case of o.Status == Rejected
+	updated.Status = protocols.Approved
+
+	return updated
+}
+
+func (o Objective) Reject() Objective {
+	updated := o.clone()
+	updated.Status = protocols.Rejected
+	return updated
+}
+
+// Update receives an ObjectiveEvent, applies all applicable event data to the DirectDefundingObjective,
+// and returns the updated objective
+func (o Objective) Update(event protocols.ObjectiveEvent) (Objective, error) {
+	if o.Id() != event.ObjectiveId {
+		return o, fmt.Errorf("event and objective Ids do not match: %s and %s respectively", string(event.ObjectiveId), string(o.Id()))
+	}
+
+	if len(event.SignedStates) > 0 {
+		for _, ss := range event.SignedStates {
+			if !ss.State().IsFinal {
+				return o, errors.New("direct defund objective can only be updated with final states")
+			}
+		}
+	}
+
+	updated := o.clone()
+	updated.C.AddSignedStates(event.SignedStates)
+
+	return updated, nil
+}
+
+// Crank inspects the extended state and declares a list of Effects to be executed
+func (o Objective) Crank(secretKey *[]byte) (Objective, protocols.SideEffects, protocols.WaitingFor, error) {
+	updated := o.clone()
+
+	sideEffects := protocols.SideEffects{}
+
+	// Input validation
+	if updated.Status != protocols.Approved {
+		return updated, sideEffects, WaitingForNothing, ErrNotApproved
+	}
+
+	latestSignedState, err := updated.C.LatestSignedState()
+	if err != nil {
+		return updated, sideEffects, WaitingForNothing, errors.New("The channel must contain at least one signed state to crank the defund objective")
+	}
+
+	// Sign a final state if it has not been signed
```

Comment:
Maybe something like "Finalize and sign a state if no supported, finalized state exists"?

Reply:
Updated in 8288b16.